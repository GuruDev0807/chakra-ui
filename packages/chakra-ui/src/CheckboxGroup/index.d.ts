import * as React from "react";
import * as StyledSystem from "styled-system";
import { BoxProps } from "../Box";

export interface ICheckboxGroup {
  id?: string;
  name?: string;
  children?: React.ReactNode;
  defaultValue?: Array<any>;
  value?: Array<any>;
  onChange?: (values?: Array<any>) => void;
  spacing?: StyledSystem.MarginProps;
  isInline?: boolean;
}

export type CheckboxGroupProps = ICheckboxGroup & BoxProps;
declare const CheckboxGroup: React.FC<CheckboxGroupProps>;
export default CheckboxGroup;
