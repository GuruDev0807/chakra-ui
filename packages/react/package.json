{
  "name": "@chakra-ui/react",
  "version": "1.7.0",
  "description": "Responsive and accessible React UI components built with React and Emotion",
  "main": "dist/chakra-ui-react.cjs.js",
  "module": "dist/chakra-ui-react.esm.js",
  "types": "dist/chakra-ui-react.cjs.d.ts",
  "files": [
    "dist"
  ],
  "sideEffects": false,
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@chakra-ui/accordion": "1.4.0",
    "@chakra-ui/alert": "1.3.0",
    "@chakra-ui/avatar": "1.3.0",
    "@chakra-ui/breadcrumb": "1.3.0",
    "@chakra-ui/button": "1.5.0",
    "@chakra-ui/checkbox": "1.6.0",
    "@chakra-ui/close-button": "1.2.0",
    "@chakra-ui/control-box": "1.1.0",
    "@chakra-ui/counter": "1.2.0",
    "@chakra-ui/css-reset": "1.1.0",
    "@chakra-ui/editable": "1.3.0",
    "@chakra-ui/form-control": "1.5.0",
    "@chakra-ui/hooks": "1.7.0",
    "@chakra-ui/icon": "1.2.0",
    "@chakra-ui/image": "1.1.0",
    "@chakra-ui/input": "1.3.0",
    "@chakra-ui/layout": "1.5.0",
    "@chakra-ui/live-region": "1.1.0",
    "@chakra-ui/media-query": "1.2.0",
    "@chakra-ui/menu": "1.8.0",
    "@chakra-ui/modal": "1.10.0",
    "@chakra-ui/number-input": "1.3.0",
    "@chakra-ui/pin-input": "1.7.0",
    "@chakra-ui/popover": "1.10.0",
    "@chakra-ui/popper": "2.4.0",
    "@chakra-ui/portal": "1.3.0",
    "@chakra-ui/progress": "1.2.0",
    "@chakra-ui/provider": "1.7.0",
    "@chakra-ui/radio": "1.4.0",
    "@chakra-ui/react-env": "1.1.0",
    "@chakra-ui/select": "1.2.0",
    "@chakra-ui/skeleton": "1.2.0",
    "@chakra-ui/slider": "1.5.0",
    "@chakra-ui/spinner": "1.2.0",
    "@chakra-ui/stat": "1.2.0",
    "@chakra-ui/switch": "1.3.0",
    "@chakra-ui/system": "1.8.0",
    "@chakra-ui/table": "1.3.0",
    "@chakra-ui/tabs": "1.6.0",
    "@chakra-ui/tag": "1.2.0",
    "@chakra-ui/textarea": "1.2.0",
    "@chakra-ui/theme": "1.12.0",
    "@chakra-ui/toast": "1.4.0",
    "@chakra-ui/tooltip": "1.4.0",
    "@chakra-ui/transition": "1.4.0",
    "@chakra-ui/utils": "1.9.0",
    "@chakra-ui/visually-hidden": "1.1.0"
  },
  "peerDependencies": {
    "@emotion/react": "^11.0.0",
    "@emotion/styled": "^11.0.0",
    "framer-motion": "3.x || 4.x || 5.x",
    "react": ">=16.8.6",
    "react-dom": ">=16.8.6"
  },
  "author": "Segun Adebayo <sage@adebayosegun.com>",
  "license": "MIT",
  "homepage": "https://chakra-ui.com/",
  "repository": {
    "type": "git",
    "url": "https://github.com/chakra-ui/chakra-ui",
    "directory": "packages/react"
  },
  "keywords": [
    "react",
    "ui",
    "design-system",
    "react-components",
    "uikit",
    "accessible",
    "components",
    "emotion",
    "library",
    "design-system"
  ],
  "storybook": {
    "title": "Chakra UI",
    "url": "https://chakra-ui.netlify.app"
  },
  "devDependencies": {
    "@emotion/react": "^11.4.0",
    "@emotion/styled": "^11.3.0",
    "framer-motion": "^4.0.0",
    "react": "^17.0.1",
    "react-dom": "^17.0.2"
  }
}
