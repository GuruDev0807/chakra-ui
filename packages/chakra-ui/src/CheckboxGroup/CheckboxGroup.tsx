/** @jsx jsx */
import { jsx } from "@emotion/core";
import {
  Children,
  cloneElement,
  useState,
  useRef,
  isValidElement,
  FormEvent,
} from "react";
import { useUID as useId } from "react-uid";
import { Box } from "../Box";
import * as React from "react";
import * as StyledSystem from "styled-system";
import { BoxProps } from "../Box";
import { CheckboxOptions } from "../Checkbox";
import { Merge } from "@chakra-ui/utils";

export interface CheckboxGroupOptions {
  /**
   * The id of the checkbox group.
   */
  id?: CheckboxOptions["id"];
  /**
   * The name of the checkbox group. This prop is passed to each checbox
   */
  name?: CheckboxOptions["name"];
  /**
   * The content of the checkbox group. Must be the `Checkbox` component
   */
  children?: React.ReactNode;
  /**
   * The initial value of the checkbox group
   */
  defaultValue?: Array<string | number>;
  /**
   * The value of the checkbox group
   */
  value?: Array<string | number>;
  /**
   * The callback fired when any children Checkbox is checked or unchecked
   */
  onChange?: (value: Array<string | number>) => void;
  /**
   * The space between each checkbox
   */
  spacing?: StyledSystem.MarginProps["margin"];
  /**
   * If `true`, the checkboxes will aligned horizontally.
   */
  isInline?: boolean;
  /**
   * The color of the checkbox when it's checked.
   */
  variantColor?: CheckboxOptions["variantColor"];
  /**
   * The size of the checkbox. It's forwarded to all children checkbox
   */
  size?: CheckboxOptions["size"];
}

export type CheckboxGroupProps = Merge<BoxProps, CheckboxGroupOptions>;

const CheckboxGroup = ({
  onChange,
  name,
  variantColor,
  size,
  defaultValue,
  isInline,
  value: valueProp,
  spacing = 2,
  children,
  ...rest
}: CheckboxGroupOptions) => {
  const [values, setValues] = useState<CheckboxGroupOptions["defaultValue"]>(
    defaultValue || [],
  );

  const { current: isControlled } = useRef(valueProp != null);
  const _values = isControlled ? valueProp : values;

  const _onChange = ({
    checked,
    value,
  }: {
    checked: boolean;
    value: string | number;
  }) => {
    if (_values == null) return;

    let newValues: any[];
    if (checked) {
      newValues = [..._values, value];
    } else {
      newValues = _values.filter(val => val != value);
    }

    !isControlled && setValues(newValues);
    onChange && onChange(newValues);
  };

  // If no name is passed, we'll generate a random, unique name
  const fallbackName = `checkbox-${useId()}`;
  const _name = name || fallbackName;

  const clones = Children.map(children, (child, index) => {
    if (!isValidElement(child)) return;

    const isLastCheckbox = React.Children.count(children) === index + 1;
    const spacingProps = isInline ? { mr: spacing } : { mb: spacing };

    return (
      <Box
        display={isInline ? "inline-block" : "block"}
        {...(!isLastCheckbox && spacingProps)}
      >
        {cloneElement(child, {
          size: size,
          variantColor: variantColor,
          name: `${_name}-${index}`,
          onChange: () =>
            _onChange({
              checked: _values ? !_values.includes(child.props.value) : false,
              value: child.props.value,
            }),
          isChecked: _values ? _values.includes(child.props.value) : false,
        })}
      </Box>
    );
  });

  return (
    <Box role="group" {...rest}>
      {clones}
    </Box>
  );
};

export default CheckboxGroup;
