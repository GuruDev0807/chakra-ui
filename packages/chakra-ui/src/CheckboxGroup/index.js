/** @jsx jsx */
import { jsx } from "@emotion/core";
import { useId } from "@reach/auto-id";
import { cloneElement, useRef, useState } from "react";
import Box from "../Box";
import { cleanChildren } from "../utils";

const CheckboxGroup = ({
  onChange,
  name,
  variantColor,
  size,
  defaultValue,
  isInline,
  value: valueProp,
  spacing = 2,
  children,
  ...rest
}) => {
  const [values, setValues] = useState(defaultValue || []);

  const { current: isControlled } = useRef(valueProp != null);
  const _values = isControlled ? valueProp : values;

  const _onChange = event => {
    const { checked, value } = event.target;
    let newValues;
    if (checked) {
      newValues = [..._values, value];
    } else {
      newValues = _values.filter(val => val !== value);
    }

    !isControlled && setValues(newValues);
    onChange && onChange(newValues);
  };

  // If no name is passed, we'll generate a random, unique name
  const fallbackName = `checkbox-${useId()}`;
  const _name = name || fallbackName;

  const validChildren = cleanChildren(children);

  const clones = validChildren.map((child, index) => {
    const isLastCheckbox = validChildren.length === index + 1;
    const spacingProps = isInline ? { mr: spacing } : { mb: spacing };

    return (
      <Box
        key={index}
        display={isInline ? "inline-block" : "block"}
        {...(!isLastCheckbox && spacingProps)}
      >
        {cloneElement(child, {
          size: size,
          variantColor: variantColor,
          name: `${_name}-${index}`,
          onChange: _onChange,
          isChecked: _values.includes(child.props.value),
        })}
      </Box>
    );
  });

  return (
    <Box role="group" {...rest}>
      {clones}
    </Box>
  );
};

export default CheckboxGroup;
