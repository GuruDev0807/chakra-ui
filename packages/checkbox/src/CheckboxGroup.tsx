import { ThemingProps } from "@chakra-ui/system"
import { createContext, __DEV__ } from "@chakra-ui/utils"
import * as React from "react"
import { useMemo } from "react"
import {
  useCheckboxGroup,
  UseCheckboxGroupProps,
  UseCheckboxGroupReturn,
} from "./CheckboxGroup.hook"

export type CheckboxGroupProps = UseCheckboxGroupProps &
  Omit<ThemingProps, "orientation"> & { children?: React.ReactNode }

export type CheckboxGroupContext = Pick<
  UseCheckboxGroupReturn,
  "onChange" | "value"
> &
  Omit<ThemingProps, "orientation">

const [CheckboxGroupCtxProvider, useCheckboxGroupCtx] = createContext<
  CheckboxGroupContext
>({
  name: "CheckboxGroupContext",
  strict: false,
})

export { useCheckboxGroupCtx }

/**
 * CheckboxGroup
 *
 * Used for multiple checkboxes which are bound in one group,
 * and it indicates whether one or more options are selected.
 *
 * @see Docs https://chakra-ui.com/checkbox
 *
 */
export const CheckboxGroup = (props: CheckboxGroupProps) => {
  const { colorScheme, size, variant, children } = props
  const { value, onChange } = useCheckboxGroup(props)

  const group = useMemo(
    () => ({
      size,
      onChange,
      colorScheme,
      value,
      variant,
    }),
    [size, onChange, colorScheme, value, variant],
  )

  return (
    <CheckboxGroupCtxProvider value={group}>
      {children}
    </CheckboxGroupCtxProvider>
  )
}

if (__DEV__) {
  CheckboxGroup.displayName = "CheckboxGroup"
}
