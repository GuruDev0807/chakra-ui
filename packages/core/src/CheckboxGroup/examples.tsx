/** @jsx jsx */
import { jsx } from "@emotion/core";
import { storiesOf } from "@storybook/react";
import { CheckboxGroup } from ".";
import { Checkbox } from "../Checkbox";
import { ThemeProvider, CSSReset } from "@chakra-ui/theme";

const stories = storiesOf("CheckboxGroup", module);

stories.addDecorator(story => {
  return (
    <ThemeProvider>
      <CSSReset />
      {story()}
    </ThemeProvider>
  );
});

stories.add("Default", () => (
  <CheckboxGroup
    variantColor="green"
    defaultValue={[1]}
    onChange={val => console.log(val)}
  >
    <Checkbox value={1}>One</Checkbox>
    <Checkbox value={2}>Two</Checkbox>
    <Checkbox value={3}>Three</Checkbox>
  </CheckboxGroup>
));
