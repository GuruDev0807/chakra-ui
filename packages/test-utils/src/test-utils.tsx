import { ChakraProvider } from "@chakra-ui/core"
import "@testing-library/jest-dom/extend-expect"
import {
  render as RTLRender,
  RenderOptions,
  fireEvent,
  RenderResult,
  Queries,
} from "@testing-library/react"
import * as React from "react"
import { toHaveNoViolations } from "jest-axe"
import serializer from "jest-emotion"

expect.addSnapshotSerializer(serializer)
expect.extend(toHaveNoViolations)

const AllProviders: React.FC = ({ children }) => (
  <ChakraProvider resetCSS>{children}</ChakraProvider>
)

export function render(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">,
): RenderResult

export function render(ui: React.ReactElement, options?: RenderOptions) {
  return RTLRender(ui, { wrapper: AllProviders, ...options })
}

export * from "@testing-library/react"

export {
  act as invoke,
  renderHook,
  RenderHookOptions,
  RenderHookResult,
} from "@testing-library/react-hooks"

export { default as userEvent } from "@testing-library/user-event"

export * from "jest-axe"

export const escape = (ui: HTMLElement) =>
  fireEvent.keyDown(ui, { key: "Escape", keyCode: 27 })
