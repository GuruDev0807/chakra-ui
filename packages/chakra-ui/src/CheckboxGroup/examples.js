/** @jsx jsx */
import { jsx } from "@emotion/core";
import { storiesOf } from "@storybook/react";
import CheckboxGroup from ".";
import Checkbox from "../Checkbox";

const stories = storiesOf("CheckboxGroup", module);
stories.add("Default", () => (
  <CheckboxGroup variantColor="green" defaultValue={["two"]}>
    <Checkbox value="one">One</Checkbox>
    <Checkbox value="two">Two</Checkbox>
    <Checkbox value="three">Three</Checkbox>
  </CheckboxGroup>
));
