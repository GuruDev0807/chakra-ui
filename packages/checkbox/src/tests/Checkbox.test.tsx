import * as React from "react"
import { render, userEvent, renderHook, invoke } from "@chakra-ui/test-utils"
import { Checkbox, useCheckbox, CheckboxHookProps } from ".."
import { CheckboxGroup } from "../CheckboxGroup"

test("Checkbox renders correctly", () => {
  const utils = render(<Checkbox />)
  expect(utils.asFragment()).toMatchSnapshot()
})

test("useCheckbox should return object", () => {
  const { result } = renderHook(() => useCheckbox({}))
  expect(typeof result.current).toBe("object")
})

test("useCheckbox should return object with 4 keys", () => {
  const { result } = renderHook(() => useCheckbox({}))
  expect(Object.keys(result.current).length).toEqual(4)
})

test("Checkbox renders correctly", () => {
  const utils = render(<Checkbox>This is custom checkbox</Checkbox>)
  expect(utils.asFragment()).toMatchSnapshot()
})

test("Uncontrolled - should check and uncheck", () => {
  const Component = () => {
    const { htmlProps, getInputProps, getCheckboxProps } = useCheckbox()

    return (
      <label {...htmlProps}>
        <input data-testid="input" {...getInputProps()} />
        <div {...getCheckboxProps()}>Checkbox</div>
      </label>
    )
  }
  const utils = render(<Component />)

  const input = utils.getByTestId("input")
  const checkbox = utils.getByText("Checkbox")

  // click the first time, it's checked
  userEvent.click(checkbox)
  expect(input).toBeChecked()
  expect(checkbox).toHaveAttribute("data-checked")

  // click the second time, it's unchecked
  userEvent.click(checkbox)
  expect(input).not.toBeChecked()
  expect(checkbox).not.toHaveAttribute("data-checked")
})

test("Uncontrolled - should not check if disabled", () => {
  const Component = () => {
    const { htmlProps, getInputProps, getCheckboxProps } = useCheckbox({
      isDisabled: true,
    })

    return (
      <label {...htmlProps}>
        <input data-testid="input" {...getInputProps()} />
        <div {...getCheckboxProps()}>Checkbox</div>
      </label>
    )
  }
  const utils = render(<Component />)

  const input = utils.getByTestId("input")
  const checkbox = utils.getByText("Checkbox")

  expect(input).toBeDisabled()
  expect(checkbox).toHaveAttribute("data-disabled")

  userEvent.click(checkbox)

  expect(input).not.toBeChecked()
  expect(checkbox).not.toHaveAttribute("data-checked")
})

test("indeterminate state", () => {
  const Component = () => {
    const { htmlProps, getInputProps, getCheckboxProps } = useCheckbox({
      isIndeterminate: true,
    })

    return (
      <label {...htmlProps}>
        <input data-testid="input" {...getInputProps()} />
        <div {...getCheckboxProps()}>Checkbox</div>
      </label>
    )
  }
  const utils = render(<Component />)

  const checkbox = utils.getByText("Checkbox")
  expect(checkbox).toHaveAttribute("data-mixed")
})

test("Controlled - should check and uncheck", () => {
  let checked = false
  const onChange = jest.fn(e => (checked = e.target.checked))

  const Component = (props: CheckboxHookProps) => {
    const { htmlProps, getInputProps, getCheckboxProps } = useCheckbox(props)

    return (
      <label {...htmlProps}>
        <input data-testid="input" {...getInputProps()} />
        <div {...getCheckboxProps()}>Checkbox</div>
      </label>
    )
  }

  const utils = render(<Component isChecked={checked} onChange={onChange} />)

  const input = utils.getByTestId("input")
  const checkbox = utils.getByText("Checkbox")

  expect(input).not.toHaveAttribute("data-checked")

  userEvent.click(checkbox)
  expect(onChange).toHaveBeenCalled()

  utils.rerender(<Component isChecked={checked} onChange={onChange} />)

  expect(onChange).toHaveBeenCalled()
  expect(checkbox).toHaveAttribute("data-checked")

  userEvent.click(checkbox)
  expect(onChange).toHaveBeenCalled()

  utils.rerender(<Component isChecked={checked} onChange={onChange} />)

  expect(onChange).toHaveBeenCalled()
  expect(checkbox).not.toHaveAttribute("data-checked")
})
