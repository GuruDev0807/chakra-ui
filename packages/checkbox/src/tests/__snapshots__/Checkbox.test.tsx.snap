// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
<DocumentFragment>
  <label
    class="chakra-checkbox css-1elxtiq"
  >
    <input
      class="chakra-checkbox__input"
      style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
      type="checkbox"
      value=""
    />
    <div
      aria-hidden="true"
      class="chakra-checkbox__control css-r1uwzp"
    />
  </label>
</DocumentFragment>
`;

exports[`Checkbox renders correctly 2`] = `
<DocumentFragment>
  <label
    class="chakra-checkbox css-1elxtiq"
  >
    <input
      class="chakra-checkbox__input"
      style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
      type="checkbox"
      value=""
    />
    <div
      aria-hidden="true"
      class="chakra-checkbox__control css-r1uwzp"
    />
    <div
      class="chakra-checkbox__label css-k4agny"
    >
      This is custom checkbox
    </div>
  </label>
</DocumentFragment>
`;
